### PR TITLE
[SPARK-21882][CORE] OutputMetrics doesn't count written bytes correctly in the saveAsHadoopDataset function

### PR DESCRIPTION
**Pull request by awarrior**
[SPARK-21882](https://issues.apache.org/jira/browse/SPARK-21882)

## What changes were proposed in this pull request?

The first job called from saveAsHadoopDataset, running in each executor, does not calculate the writtenBytes of OutputMetrics correctly (writtenBytes is 0). The reason is that we did not initialize the callback function called to find bytes written in the right way. As usual, statisticsTable which records statistics in a FileSystem must be initialized at the beginning (this will be triggered when open SparkHadoopWriter). One solution for this issue is to switch the initialization order of HadoopOutputMetrics and SparkHadoopWriter.

## How was this patch tested?

Existing tests

### CODE REVIEW

**Comment by jiangxb1987** [comment]
ok to test

**Comment by jiangxb1987** [comment]
This should be a valid fix. Could you add a test case for this?

**Comment by gatorsmile** [comment]
ok to test

**Comment by awarrior** [comment]
@jiangxb1987 ok. I add one later.

**Comment by awarrior** [comment]
I met a trouble when I write a test case. It seems that this issue won't be triggered in only one node. I found that Driver node do createPathFromString so that there is no problem.

>   java.lang.Thread.State: RUNNABLE
	  at org.apache.hadoop.fs.FileSystem.getStatistics(FileSystem.java:3271)
	  - locked <0x1211> (a java.lang.Class)
	  at org.apache.hadoop.fs.FileSystem.initialize(FileSystem.java:202)
	  at org.apache.hadoop.fs.RawLocalFileSystem.initialize(RawLocalFileSystem.java:92)
	  at org.apache.hadoop.fs.LocalFileSystem.initialize(LocalFileSystem.java:47)
	  at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2598)
	  at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:91)
	  at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2632)
	  at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2614)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:370)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:169)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:354)
	  at org.apache.hadoop.fs.Path.getFileSystem(Path.java:296)
	  at org.apache.spark.internal.io.SparkHadoopWriterUtils$.createPathFromString(SparkHadoopWriterUtils.scala:55)

Does anyone know how to test in this case?

**Comment by jiangxb1987** [comment]
You don't have to use `saveAsHadoopFile`, just call `saveAsHadoopDataset` directly?

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala [hunk 1]
```diff
@@ -112,11 +112,12 @@ object SparkHadoopWriter extends Logging {
       jobTrackerId, sparkStageId, sparkPartitionId, sparkAttemptNumber)
     committer.setupTask(taskContext)
 
-    val (outputMetrics, callback) = initHadoopOutputMetrics(context)
-
     // Initiate the writer.
     config.initWriter(taskContext, sparkPartitionId)
     var recordsWritten = 0L
+    
+    // Initialize callback function after the writer.
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala]: The comment is better to explain why we must initialize the callback here. E.g., `We must initialize the callback for calculating bytes written after the statistic table is initialized in FileSystem`.

Comment (awarrior) [reply core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala]: ok

**Comment by awarrior** [comment]
@jiangxb1987 well, I passed that part above but met other initialization chances before runJob. They are in the write function of SparkHadoopWriter.

>     
    // Assert the output format/key/value class is set in JobConf.
    config.assertConf(jobContext, rdd.conf)                //// <= chance

    val committer = config.createCommitter(stageId)
    committer.setupJob(jobContext)                //// <= chance

    // Try to write all RDD partitions as a Hadoop OutputFormat.
    try {
      val ret = sparkContext.runJob(rdd, (context: TaskContext, iter: Iterator[(K, V)]) => {
        executeTask(
          context = context,
          config = config,
          jobTrackerId = jobTrackerId,
          sparkStageId = context.stageId,
          sparkPartitionId = context.partitionId,
          sparkAttemptNumber = context.attemptNumber,
          committer = committer,
          iterator = iter)
      })

One trace list:

> java.lang.Thread.State: RUNNABLE
	  at org.apache.hadoop.fs.FileSystem.getStatistics(FileSystem.java:3270)
	  - locked <0x126a> (a java.lang.Class)
	  at org.apache.hadoop.fs.FileSystem.initialize(FileSystem.java:202)
	  at org.apache.hadoop.fs.RawLocalFileSystem.initialize(RawLocalFileSystem.java:92)
	  at org.apache.hadoop.fs.LocalFileSystem.initialize(LocalFileSystem.java:47)
	  at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2598)
	  at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:91)
	  at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2632)
	  at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2614)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:370)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:169)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:354)
	  at org.apache.hadoop.fs.Path.getFileSystem(Path.java:296)
	  at org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter.<init>(FileOutputCommitter.java:91)
	  at org.apache.hadoop.mapred.FileOutputCommitter.getWrapped(FileOutputCommitter.java:65)
	  at org.apache.hadoop.mapred.FileOutputCommitter.setupJob(FileOutputCommitter.java:131)
	  at org.apache.hadoop.mapred.OutputCommitter.setupJob(OutputCommitter.java:233)
	  at org.apache.spark.internal.io.HadoopMapReduceCommitProtocol.setupJob(HadoopMapReduceCommitProtocol.scala:125)
	  at org.apache.spark.internal.io.SparkHadoopWriter$.write(SparkHadoopWriter.scala:74)

**Comment by jiangxb1987** [comment]
ping @awarrior Could you please fix the test failure?

**Comment by awarrior** [comment]
@jiangxb1987 well, the test case is hard to construct if we just run app in local like comments above. Any ideas to crack?

**Comment by mridulm** [comment]
The change looks good to me, pending test case; thanks for looking into this @awarrior

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
retest this please

**Review by srowen — CHANGES_REQUESTED** [review]
@awarrior if you'll fix the style issue (line too long I think) we can probably merge this

**Comment by gatorsmile** [comment]
@Ngone51 Could you please take this over?

**Comment by Ngone51** [comment]
sure @gatorsmile

**Comment by srowen** [comment]
@Ngone51 did you open another PR? I can try if not.

**Comment by Ngone51** [comment]
Sorry, haven't. Please try it, thanks. @srowen

**Comment by srowen** [comment]
Moved to https://github.com/apache/spark/pull/24863

**Pull request closed without merging**